File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/Status.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Git Modification Status
 * <p>
 * see "diff-raw status letters" at https://github.com/git/git/blob/35f6318d44379452d8d33e880d8df0267b4a0cd0/diff.h#L326
 */
enum Status {
    ADDED('A'),
    COPIED('C'),
    DELETED('D'),
    MODIFIED('M'),
    RENAMED('R'),
    TYPE_CHANGED('T'),
    UNKNOWN('X'),
    UNMERGED('U');

    public static final List<Character> ALL_STATUS_LETTERS =
            Stream.of(Status.class.getEnumConstants())
                    .map(Status::statusLetter)
                    .collect(Collectors.toList());

    private final char letter;

    Status(char t) {
        letter = t;
    }

    public static Status byCharacter(char c) {
        return Stream.of(Status.class.getEnumConstants())
                .filter(status -> status.letter == c)
                .findFirst()
                .orElse(UNKNOWN);
    }

    public Modification.Type toModificationType() {
        switch (this) {
            case ADDED:
                return Modification.Type.ADD;
            case DELETED:
                return Modification.Type.DELETE;
            case MODIFIED:
                return Modification.Type.MODIFY;
            case RENAMED:
                return Modification.Type.RENAME;
            default:
                return Modification.Type.UNKNOWN;
        }
    }

    public char statusLetter() {
        return letter;
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/Status.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification

/**
 * Git Modification Status
 *
 *
 * see "diff-raw status letters" at https://github.com/git/git/blob/35f6318d44379452d8d33e880d8df0267b4a0cd0/diff.h#L326
 */
internal enum class Status(private val letter: Char) {

    ADDED('A'),
    COPIED('C'),
    DELETED('D'),
    MODIFIED('M'),
    RENAMED('R'),
    TYPE_CHANGED('T'),
    UNKNOWN('X'),
    UNMERGED('U');

    fun toModificationType(): Modification.Type {
        return when (this) {
            ADDED -> Modification.Type.ADD
            DELETED -> Modification.Type.DELETE
            MODIFIED -> Modification.Type.MODIFY
            RENAMED -> Modification.Type.RENAME
            else -> Modification.Type.UNKNOWN
        }
    }

    fun statusLetter(): Char {
        return letter
    }

    companion object {

        val ALL_STATUS_LETTERS: List<Char> = Status.values().map { it.statusLetter() } //sanity check

        fun byCharacter(c: Char): Status {
            return Status.values().firstOrNull { status -> status.letter == c } ?: UNKNOWN
        }
    }
}
